### ISSUE TITLE
[Bug]: download complete but cant create library item (ios)

### ISSUE THREAD

**Issue by xanthos84** [advplyr/audiobookshelf-app#961]
### Describe the issue

downloading items does not work on last version
ios 0.9.67 / 0.9.68

### Steps to reproduce the issue

1. open app
2. download item for offline listening
3. items download progresses until 100%
4. error msg „download complete but cant create library item“

### Audiobookshelf version

2.5 server 0.9.67/0.9.68 ios

### How are you running audiobookshelf?

Docker

**Comment by xanthos84** [comment]
> Are you able to open the browser and download successfully from the website?

yes that works fine for single files and for complete packages (zip).

**Comment by xanthos84** [comment]
this is still an issue with the newest app on iOS + newest server version!

**Comment by Jeor** [comment]
Having same issue

Server 2.8.1
ios (self signed) 0.9.73-beta

**Comment by soberhofer** [comment]
Having the same issue. App is signed and installed using a valid dev certificate. How can I further troubleshoot this?

**Comment by jerry00020001** [comment]
Also getting this error, only happens on my wife's phone, any book. Tried hostname, then IP. Tried deleting app and re-installing. Tried creating a second account, issue still persists. Tried restarting iPhone. re-scanning book. works fine on my phone.

**Comment by clincher61** [comment]
Also getting this error. Client version 0.9.74.  Server version 2.9.0.  Every attempt to download a book yields "Item download complete but failed to create library item" error message on the client.  App is personally signed and sideloaded (not part of the test flight beta).  iOS 17.4.1.  

I suspect this may have something to do with personally signed apps as it sounds like everyone reporting above is not in the Test Flight beta.  ?

**Comment by cofifield** [comment]
**TLDR: Use AltStore to sideload the app.**

I have found the root cause of this problem. It is occurring due to a missing sandbox entitlement which seems to be coming from how users are choosing to sideload their app. It is **not** an application bug/build/configuration issue. At this time I have tested Sideloadly and AltStore with sideloading via Altstore being the solution. Other sideloading methods may work but I have not tested them.

When building the application from source and deploying to a simulator or physical device, we can review the console logs while attempting to download a file from the server. A proper download looks like this:
`⚡️  [log] - Local library item {"serverConnectionConfigId":"aHR0cHM6Ly9hYi5jZi1sLmNvbUBjb2ZpZmllbGQ=","localFiles":[{"size":482114117,"contentUrl":"file:///var/mobile/Containers/Data/Application/608A8FB7-6BA5-4393-9761-92B505875F5B/Documents/9550e1ac-ebc0-4b6e-b233-8a563aa379be/Harry%20Potter%20and%20the%20Philosopher's%20Stone%20-%20Harry%20Potter%20and%20the%20Philosopher's%20Stone.m4b","id":"9550e1ac-ebc0-4b6e-b233-8a563aa379be_SGFycnkgUG90dGVyIGFuZCB0aGUgUGhpbG9zb3BoZXIncyBTdG9uZSAtIEhhcnJ5IFBvdHRlciBhbmQgdGhlIFBoaWxvc29waGVyJ3MgU3RvbmUubTRi","filename":"Harry Potter and the Philosopher's Stone - Harry Potter and the Philosopher's Stone.m4b","basePath":"Harry Potter and the Philosopher's Stone - Harry Potter and the Philosopher's Stone.m4b","mimeType":"audio/mp4"},{"mimeType":"image/jpg","contentUrl":"file:///var/mobile/Containers/Data/Application/608A8FB7-6BA5-4393-9761-92B505875F5B/Documents/9550e1ac-ebc0-4b6e-b233-8a563aa379be/cover.jpg","id":"9550e1ac-ebc0-4b6e-b233-8a563aa379be_Y292ZXIuanBn","filename":"cover.jpg","basePath":"cover.jpg","size":32016}],"serverAddress":"https://your-server-here.com","isLocal":true,"libraryItemId":"9550e1ac-ebc0-4b6e-b233-8a563aa379be","contentUrl":"file:///var/mobile/Containers/Data/Application/608A8FB7-6BA5-4393-9761-92B505875F5B/Documents/9550e1ac-ebc0-4b6e-b233-8a563aa379be/","media":{"audioFiles":[{"index":1,"metadata":{"filename":"Harry Potter and the Philosopher's Stone - Harry Potter and the Philosopher's Stone.m4b","ext":".m4b","relPath":"Harry Potter and the Philosopher's Stone - Harry Potter and the Philosopher's Stone.m4b","path":"/audiobooks/J.K. Rowling/Harry Potter/1 - Harry Potter and the Philosopher's Stone/Harry Potter and the Philosopher's Stone - Harry Potter and the Philosopher's Stone.m4b","size":482114117},"ino":"36522"}],"tags":[],"ebookFile":null,"metadata":{"explicit":false,"authors":[{"name":"J.K. Rowling","id":"e414a5ac-0cc7-42b9-b410-ac6c8d609f7c","coverPath":null}],"isbn":null,"publishedYear":"1997","genres":["Academy","Portal Fantasy (YA)"],"language":"English","authorName":"J.K. Rowling","publisher":"Pottermore Publishing","narratorName":"Stephen Fry","subtitle":null,"asin":"B079TM3YHV","title":"01 Harry Potter and the Philosopher's Stone","authorNameLF":"Rowling, J.K.","publishedDate":null,"narrators":["Stephen Fry"],"feedUrl":null,"seriesName":"Harry Potter #1","author":null,"description":"Unabridged"},"size":482114117,"episodes":[],"autoDownloadEpisodes":null,"libraryItemId":"9550e1ac-ebc0-4b6e-b233-8a563aa379be","chapters":[{"end":28.813991,"id":0,"title":"Opening Credits","start":0},{"end":1928.717982,"id":1,"title":"Chapter 1: The Boy Who Lived","start":28.813991},{"end":3258.663968,"id":2,"title":"Chapter 2: The Vanishing Glass","start":1928.717982},{"end":4775.994966,"id":3,"title":"Chapter 3: The Letters from No One","start":3258.663968},{"end":6246.165964,"id":4,"title":"Chapter 4: The Keeper of the Keys","start":4775.994966},{"end":8890.128957,"id":5,"title":"Chapter 5: Diagon Alley","start":6246.165964},{"end":11176.736939,"id":6,"title":"Chapter 6: The Journey from Platform Nine and Three-Quarters","start":8890.128957},{"end":12950.763923,"id":7,"title":"Chapter 7: The Sorting Hat","start":11176.736939},{"end":14076.652902,"id":8,"title":"Chapter 8: The Potions Master","start":12950.763923},{"end":15918.2739,"id":9,"title":"Chapter 9: The Midnight Duel","start":14076.652902},{"end":17490.101882,"id":10,"title":"Chapter 10: Hallowe'en","start":15918.2739},{"end":18700.069864,"id":11,"title":"Chapter 11: Quidditch","start":17490.101882},{"end":20766.296848,"id":12,"title":"Chapter 12: The Mirror of Erised","start":18700.069864},{"end":21985.808844,"id":13,"title":"Chapter 13: Nicolas Flamel","start":20766.296848},{"end":23265.900839,"id":14,"title":"Chapter 14: Norbert the Norwegian","start":21985.808844},{"end":25329.318821,"id":15,"title":"Chapter 15: The Forbidden Forest","start":23265.900839},{"end":27916.578821,"id":16,"title":"Chapter 16: Through the Trapdoor","start":25329.318821},{"end":30243.98381,"id":17,"title"
`

The big takeaway here is that the system is able to write and access
 `/private/var/mobile/Library/Caches/com.apple.nsurlsessiond/Downloads/com.audiobookshelf.app.XXXXXX/`
while downloading the temporary file and moving it into the user's library.

The `ENABLE_USER_SCRIPT_SANDBOXING` is already set to "no" in the app's build configuration. With help from [this article](https://stackoverflow.com/questions/76590131/error-while-build-ios-app-in-xcode-sandbox-rsync-samba-13105-deny1-file-w) I was able to determine the app configuration was correct.

As seen by other users above, when the application is signed without the correct entitlements we receive a console message
`iPhone kernel(Sandbox)[0] <Error>: Sandbox: Audiobookshelf(8348) deny(1) file-write-unlink /private/var/mobile/Library/Caches/com.apple.nsurlsessiond/Downloads/com.audiobookshelf.app.XXXXXXX/CFNetworkDownload_UyO830.tmp`

Hope this helps! I believe it would be beneficial to have a notice posted as I see users potentially running into this.

**Comment by advplyr** [comment]
Thanks, can the OP confirm if they were also sideloading? @xanthos84

**Comment by xanthos84** [comment]
yes sideloading the app.
need to see if i can switch to something like altstore,
as I use my own system to sign the apps - so its quite embarrassing to use something else 3rd party to get the app running like it should.
maybe i can get a TestFlight invitation?

**Comment by xanthos84** [comment]
brilliant news! just opted in!

**Comment by Jeor** [comment]
Latest release installed via Altstore is no longer having the issue.  Previously, Sideloadly was used unsuccessfully FWIW.

**Comment by megawubs** [comment]
I also ran into this problem earlier this week. For me it happened because I didn't have enough free space on my iPhone. After cleaning a few gbs up, the download succeeded.

**Comment by I-Would-Like-To-Report-A-Bug-Please** [comment]
Have same thing. Sideloadly worked absolutely fine, but e-sign and scarlet are having issues.

**Comment by Tihersarc** [comment]
Same issue here using e-sign on v0.9.76.

**Comment by cduckett85** [comment]
Having the same issue sideloading with feather; is there a workaround if installing IPA from iOS?

Edit: I was able to get this working by signing the app with feather and then installing with Sideloadly

**Comment by I-Would-Like-To-Report-A-Bug-Please** [comment]
I was able to get by a chance of luck into Testflight so I never tried any other options, if it's still full there's Prologue Testflight too which I think had still spaces open few days ago.

**Comment by Tihersarc** [comment]
I ended up using LiveContainer through SideStore.

**Comment by 11iq** [comment]
ipa sideloaded via Feather failed. Worked via LiveContainer. Thanks @Tihersarc

**Comment by DrNgo** [comment]
Digging through this thread, there are actually **two distinct root causes** behind "download complete but can't create library item", and [#1924](https://github.com/advplyr/audiobookshelf-app/pull/1924) only addresses one of them.

**1. Network-dependent completion (fixed by #1924).** On iOS, `onItemDownloadComplete` was emitted only from inside a `getLibraryItemWithProgress` server callback, so if that call failed or returned nothing (offline, rate-limited, or expired token) the files finished downloading but never got turned into a local library item. #1924 finalizes from the already-downloaded metadata immediately, so completion no longer depends on a network round-trip.

**2. Sideloading without the correct entitlements (NOT fixed by #1924 — environmental).** As documented earlier in this thread, when the app is signed without the right entitlements iOS Sandbox denies the file move out of the `nsurlsessiond` download cache:

```
Sandbox: Audiobookshelf(...) deny(1) file-write-unlink
/private/var/mobile/Library/Caches/com.apple.nsurlsessiond/Downloads/com.audiobookshelf.app.XXXX/CFNetworkDownload_XXXX.tmp
```

That's a signing/OS issue, not app code (the app's own `ENABLE_USER_SCRIPT_SANDBOXING` is already correct). Known-good install paths: **TestFlight, AltStore, LiveContainer (SideStore)**. Reported failing in various configs: **Sideloadly, e-sign, Scarlet, Feather**. (A third, unrelated trigger for the same message is simply a **full disk** — the move fails when there's no free space.)

So if you're on TestFlight/AltStore and still hit this, #1924 should fix it. If you're sideloading with a tool that doesn't grant the entitlements, the fix is to switch install methods — no app-side change can grant those entitlements. With #1924, cause-#2 users will at least get a clear "Download failed" toast instead of a silent hang, but the download itself still can't succeed until it's installed through a method that grants the entitlements.
